src/particle.f90

```diff
@@ -0,0 +1,2187 @@
+
+  !+-------------------------------------------------------------------+
+  !| this function is to retune the size of a array                    |
+  !+-------------------------------------------------------------------+
+  !| CHANGE RECORD                                                     |
+  !| -------------                                                     |
+  !| 16-Jun-2022  | Created by J. Fang STFC Daresbury Laboratory       |
+  !+-------------------------------------------------------------------+
+  pure function size_particle(var) result(nsize)
+    !
+    type(partype),allocatable,intent(in) :: var(:)
+    integer :: nsize
+    !
+    if(allocated(var)) then
+      nsize=size(var)
+    else
+      nsize=0
+    endif
+    !
+    return
+    !
+  end function size_particle
+  !
+  pure function size_integer(var) result(nsize)
+    !
+    integer,allocatable,intent(in) :: var(:)
+    integer :: nsize
+    !
+    if(allocated(var)) then
+      nsize=size(var)
+    else
+      nsize=0
+    endif
+    !
+    return
+    !
+  end function size_integer
+  !+-------------------------------------------------------------------+
+  !| The end of the subroutine size_particle.                          |
+  !+-------------------------------------------------------------------+
+
+  !+-------------------------------------------------------------------+
+  !| this subroutine clean superfluous elements in a array             |
+  !+-------------------------------------------------------------------+
+  !| CHANGE RECORD                                                     |
+  !| -------------                                                     |
+  !| 07-Nov-2018  | Created by J. Fang STFC Daresbury Laboratory       |
+  !+-------------------------------------------------------------------+
+  subroutine mclean_mytype(var,n)
+    !
+    ! arguments
+    real(mytype),allocatable,intent(inout) :: var(:)
+    integer,intent(in) :: n
+    !
+    ! local data
+    real(mytype),allocatable :: buffer(:)
+    integer :: m
+    logical :: lefc
+    !
+    if(.not.allocated(var)) return
+    !
+    if(n<=0) then
+      deallocate(var)
+      return
+    endif
+    !
+    ! clean
+    allocate(buffer(n))
+    !
+    buffer(1:n)=var(1:n)
+    !
+    deallocate(var)
+    !
+    call move_alloc(buffer,var)
+    !
+  end subroutine mclean_mytype
+  !!
+  subroutine mclean_particles(var,n)
+    !
+    ! arguments
+    type(partype),allocatable,intent(inout) :: var(:)
+    integer,intent(in) :: n
+    !
+    ! local data
+    type(partype),allocatable :: buffer(:)
+    integer :: m
+    logical :: lefc
+    !
+    if(.not.allocated(var)) return
+    !
+    if(n<=0) then
+      deallocate(var)
+      return
+    endif
+    !
+    ! clean
+    allocate(buffer(n))
+    !
+    buffer(1:n)=var(1:n)
+    !
+    deallocate(var)
+    !
+    call move_alloc(buffer,var)
+    !
+  end subroutine mclean_particles
```

Comment:
I think you can write a single/generic version of this using `class(*)` as the type of `var`

Reply:
not work with type(partype) :: buffer if use class(*)